### PR TITLE
refactor: (GenAI) Reorganized Reasoning Engine Samples (Group С)

### PR DESCRIPTION
**Pull request by Thoughtseize1**
## Description

Created new folder:`/generative_ai/reasoning_engine`

I changed the output filenames to prevent conflicts with `extensions` folder.

| Region_tag                                              | New file                   |
|---------------------------------------------------------|----------------------------|
| generativeaionvertexai_create_reasoning_engine_basic    | create_reasoning_engine_example.py         |
| generativeaionvertexai_create_reasoning_engine_advanced | create_reasoning_engine_advanced_example.py |
| generativeaionvertexai_list_reasoning_engines           | list_reasoning_engine_example.py            |
| generativeaionvertexai_get_reasoning_engine             | get_reasoning_engine_example.py             |
| generativeaionvertexai_query_reasoning_engine           | query_reasoning_engine_example.py           |
| generativeaionvertexai_delete_reasoning_engine           | delete_reasoning_engine_example.py           |

❗Files need to be deleted after approval:

- `/generative_ai/gemini_reasoning_engine.py`
- `/generative_ai/gemini_reasoning_engine_test.py`

## Checklist
- [x] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [ ] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [x] **Tests** pass:   `nox -s py-3.9` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [x] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [ ] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    app = LangchainApp(project=PROJECT_ID, location="us-central1")
+    app.set_up()
+    print(app.query("What is Vertex AI?"))
+
+    # Create a remote app with reasoning engine
+    # This may take 1-2 minutes to finish because it builds a container and turn up HTTP servers.
+    reasoning_engine = reasoning_engines.ReasoningEngine.create(
+        LangchainApp(project=PROJECT_ID, location="us-central1"),
+        requirements=[
+            "google-cloud-aiplatform==1.50.0",
```
Comment (holtskinner) [inline generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py:80]: Maybe use a newer version of the library.

File: generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            """
+            return self.chain.invoke({"text": question}).content
+
+    # Locally test
+    app = LangchainApp(project=PROJECT_ID, location="us-central1")
+    app.set_up()
+    print(app.query("What is Vertex AI?"))
+
+    # Create a remote app with reasoning engine
+    # This may take 1-2 minutes to finish because it builds a container and turn up HTTP servers.
```
Comment (koverholt) [inline generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py:76]: There technically isn't a container build step (or at least wasn't last time I checked), so perhaps simplify this comment (here and in all code samples) to `Deployment of the app should take a few minutes to complete`.

File: generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                str: The LLM response.
+            """
+            return self.chain.invoke({"text": question}).content
+
+    # Locally test
+    app = LangchainApp(project=PROJECT_ID, location="us-central1")
+    app.set_up()
+    print(app.query("What is Vertex AI?"))
+
+    # Create a remote app with reasoning engine
```
Comment (koverholt) [inline generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py:75]: Reasoning Engine should be capitalized here and throughout.

File: generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    # This may take 1-2 minutes to finish because it builds a container and turn up HTTP servers.
+    reasoning_engine = reasoning_engines.ReasoningEngine.create(
+        LangchainApp(project=PROJECT_ID, location="us-central1"),
+        requirements=[
+            "google-cloud-aiplatform==1.50.0",
+            "langchain-google-vertexai",
+            "langchain-core",
+            "cloudpickle==3",
```
Comment (koverholt) [inline generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py:83]: In addition to Holt's comment, consider using the set of dependencies (here and in other code samples) that we use in the sample notebooks:

https://github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/reasoning-engine/intro_reasoning_engine.ipynb

```
"google-cloud-aiplatform[langchain,reasoningengine]",
"cloudpickle==3.0.0",
"pydantic==2.7.4",
```

(Plus any additional deps needed for the particular code sample, in this case there are no additional deps needed).

**Review by koverholt — APPROVED** [review]
Left a few comments. Everything else LGTM. Thanks for the updates and reorg!

Also CCing @yeesian for review (as I'm unable to add him directly as a reviewer) for these Reasoning Engine code samples used in docs.

File: generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py (hunk 2 above)
Comment (Thoughtseize1) [reply generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py:76]: Done

File: generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py (hunk 3 above)
Comment (Thoughtseize1) [reply generative_ai/reasoning_engine/create_reasoning_engine_advanced_example.py:75]: Done

**Review by msampathkumar — APPROVED** [review]
LGTM

**Pull request merged**